design/src/rfcs/rfc0034_generating_reusable_code.md

```diff
@@ -0,0 +1,232 @@
+
+## Background
+
+A basic property of Smithy is that shapes can be reused in a variety of contexts, which reduces duplication.
+
+We use the following model throughout the RFC:
+
+```smithy
+structure Food {
+    @httpLabel
+    type: String
+}
+
+@http(method: "GET", uri: "/dog/{type}")
+operation Dog {
+    input: Food
+}
+
+@http(method: "GET", uri: "/cat/{type}")
+operation Cat {
+    input: Food
+}
+
+@restJson1
+service RestJsonPetService {
+    operations: [
+        Dog,
+        Cat
+    ]
+}
+
+@awsJson1_0
+service AwsJsonPetService {
+    operations: [
+        Dog,
+        Cat
+    ]
+}
+```
+
+At high-level, the [Smithy library](https://github.com/awslabs/smithy-rs) library provides not only a way to parse the AST defined by the [Smithy IDL](https://smithy.io/2.0/spec/idl.html), but also exposes the conversion of the AST into an ASG which exposed through the [Walker](https://smithy.io/2.0/guides/building-codegen/mapping-shapes-to-languages.html#computing-a-service-closure) API.
+
+We visually represent the ASG of this model as:
+
+```mermaid
+graph TD
+    A[RestJsonPetService]
+    B[AwsJsonPetService]
+    C[Dog]
+    D[Cat]
+    E[Food]
+    A --> C
+    A --> D
+    B --> C
+    B --> D
+    C --> E
+    D --> E
+```
+
+Currently, the ASG of the _generated_ Rust code looks like this:
+
+```mermaid
+graph TD
+    A[RestJsonPetService]
+    B[AwsJsonPetService]
+    Dog1[Dog]
+    Dog2[Dog]
+    Cat1[Cat]
+    Cat2[Cat]
+    DogInput1[DogInput]
+    DogInput2[DogInput]
+    CatInput1[CatInput]
+    CatInput2[CatInput]
+    A --> Cat1
+    A --> Dog1
+    B --> Cat2
+    B --> Dog2
+    Cat1 --> CatInput1
+    Cat2 --> CatInput2
+    Dog1 --> DogInput1
+    Dog2 --> DogInput2
+```
+
+Importantly, the `Food` structure gets transformed into a four incompatible types depending the context in which it is used. An obvious incompatibility is that `RestJsonPetService > Dog > Food` is transformed to `DogInput` while `RestJsonPetService > Cat > Food` is transformed to `CatInput`. The differences go deeper, these `DogInput` and `CatInput` have incompatible `FromRequest` implementations:
+
+```rust
+trait FromRequest<P> {
+    async fn from_request(request: http::Request) -> Self;
+}
+
+impl FromRequest<RestJson1> for DogInput {
+    async fn from_request(request: http::Request) -> Self {
+        /* Deserialize using from the "/dog/{type}" URI */
+    }
+}
+
+impl FromRequest<RestJson1> for CatInput {
+    async fn from_request(request: http::Request) -> Self {
+        /* Deserialize using from the "/cat/{type}" URI */
+    }
+}
+```
+
+If we removed the name aliasing, and named the Rust type `Food` in both cases, we'd then encounter a conflicting implementation.
+
+In contrast, the `FromRequest` implementations generated from `RestJsonPetService > Dog > Food` and `AwsJsonPetService > Dog > Food` are compatible:
+
+```rust
```

Comment:
They are compatible in the sense that `impl FromRequest<RestJson1>` can co-exist with `impl FromRequest<AwsJson1_0>`. Whereas the snippet above is incompatible because they're both `impl FromRequest<RestJson1>` with non-identical implementations

Reply:
There's probably a whole section about how Rusts coherence rules play nice with all the principles. For example, if you generate for example the `OperationShape` ZST in the `operationShape` method in alignment with (3) and the a "`ServiceShape`" ZST in the `serviceShape` method then you can always impl `FromRequest<Service, Op>` without conflicts - even when you had generated the service, operation and input in separate crates.